server/mscalendar/availability.go

```diff
@@ -124,6 +128,49 @@ func (m *mscalendar) setUserStatuses(filteredUsers store.UserIndex, schedules []
 	return utils.JSONBlock(schedules), nil
 }
 
+func (m *mscalendar) notifyUpcomingEvent(mattermostUserID string, items []remote.ScheduleItem) {
+	for _, scheduleItem := range items {
+		loc, err := time.LoadLocation(scheduleItem.Start.Timezone)
+		if err != nil {
+			m.Logger.Errorf("problem loading location: %s", err.Error())
+			continue
+		}
+		now := time.Now().In(loc)
+		start, err := time.Parse("2006-01-02T15:04:05MST", scheduleItem.Start.DateTime+scheduleItem.Start.Timezone)
+		if err != nil {
+			m.Logger.Errorf("problem parsing date: %s", err.Error())
+			continue
+		}
+		if now.Add(upcomingEventNotificationTime).After(start.Add(-upcomingEventNotificationWindow)) && now.Add(upcomingEventNotificationTime).Before(start.Add(upcomingEventNotificationWindow)) {
+			m.renderScheduleItem(mattermostUserID, scheduleItem)
+		}
+	}
+}
+
+func (m *mscalendar) renderScheduleItem(mattermostUserId string, s remote.ScheduleItem) {
+	message := "You have an upcoming event:"
+	start, err := time.Parse("2006-01-02T15:04:05MST", s.Start.DateTime+s.Start.Timezone)
+	if err != nil {
+		m.Logger.Errorf("problem parsing start date: %s", err.Error())
+		return
+	}
+
+	end, err := time.Parse("2006-01-02T15:04:05MST", s.End.DateTime+s.End.Timezone)
+
+	message = fmt.Sprintf("\n%s-%s", start.Format("15:04MST"), end.Format("15:04MST"))
+	if s.Subject == "" {
+		message += fmt.Sprintf("Subject: Not available. Check your privacy settings so we can show you the subject.")
```

Comment:
I like this method of error handling :+1: Is it possible that the event simply doesn't have a subject? I'm thinking Outlook wouldn't allow that.

Reply:
Apparently, you can have an event without subject. I will change the message to something more general like:

"Subject: Not available. The event may not have a subject or there might be a problem with the permissions. If you believe there is a permissions problem, check your privacy settings so we can show you the subject."

Reply:
@aaronrothschild thoughts on my previous comment?